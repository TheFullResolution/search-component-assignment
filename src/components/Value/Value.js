// @flow

import { Component } from 'react'
import type { Node } from 'react'
import axios from 'axios'
import matchSorter from 'match-sorter'
import throttle from 'lodash.throttle'
import equal from 'fast-deep-equal'

export type AutoCompleteItem = {
  searchterm: string,
  nrResults: number
}

interface State {
  autoComplete: AutoCompleteItem[];
  error: boolean;
  suggestions: AutoCompleteItem[];
  value: string;
}

export interface ValueRenderProps extends State {
  onChange: (value: string) => void;
  resetValue: () => void;
}

type Props = {
  children: (params: ValueRenderProps) => Node
}

export class Value extends Component<Props, State> {
  state = {
    suggestions: [],
    autoComplete: [],
    error: false,
    value: ''
  }
  render() {
    return this.props.children({
      ...this.state,
      onChange: this.onChange,
      resetValue: this.resetValue
    })
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (prevState.value !== this.state.value && this.state.value.length > 2) {
      this.throttledFetchSuggestions()
    }
  }

  onChange = (value: string) => {
    this.setState(
      (prevState: State) => ({ value }),
      () => {
        if (value) {
          this.updateAutoComplete()
        }
      }
    )
  }

  resetValue = () => {
    this.setState(() => ({ value: '' }))
  }

  throttledFetchSuggestions = throttle(
    () => {
      this.fetchSuggestions()
    },
    500,
    { leading: true }
  )

  fetchSuggestions = async () => {
    const fetchedData = await fetchData(this.state.value)

    if (fetchedData.error) {
      this.setState(() => ({ error: true }))
    } else if (
      fetchedData.suggestions &&
      !equal(this.state.suggestions, fetchedData.suggestions)
    ) {
      this.setState(
        () => ({
          suggestions: fetchedData.suggestions
        }),
        () => {
          this.updateAutoComplete()
        }
      )
    }
  }

  updateAutoComplete = () => {
    const autoComplete = matchSorter(this.state.suggestions, this.state.value, {
      keys: ['searchterm']
    })

    this.setState(() => ({
      autoComplete
    }))
  }
}

type FetchDataReturn =
  | {| suggestions: AutoCompleteItem[] |}
  | {| error: boolean |}

const fetchData = (value: string): FetchDataReturn => {
  return axios
    .get(`/search?q=${value}`)
    .then(response => ({ suggestions: response.data.suggestions }))
    .catch(error => ({ error: true }))
}
