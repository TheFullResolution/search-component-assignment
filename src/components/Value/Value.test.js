// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import { Value } from './Value.js'

it('Value: default', () => {
  const component = renderer.create(
    <Value>{props => <div>{JSON.stringify(props)}</div>}</Value>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
