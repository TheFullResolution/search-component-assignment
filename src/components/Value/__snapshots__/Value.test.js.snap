// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value: default 1`] = `
<div>
  {"suggestions":[],"autoComplete":[],"error":false,"value":""}
</div>
`;
